host/bcm283x/dma.go

```diff
@@ -702,6 +760,237 @@ func startPWMbyDMA(p *Pin, rng, data uint32) (*dmaChannel, *videocore.Mem, error
+
+	// Other constants during the loop.
+	// Waits does not seem to work as expected. Not counted as DREQ pulses?
+	// Use PWM's rng1 instead for this.
+	//waits := divs - 1
+	dest := [2]uint32{
+		gpioBaseAddr + 0x28 + 4*uint32(p.number/32), // clear
+		gpioBaseAddr + 0x1C + 4*uint32(p.number/32), // set
+	}
+
+	// Render the controlBlock's to trigger the bit trigger for either Set or
+	// Clear GPIO memory registers.
+	last = getBit(bits[0], 0, msb)
+	index := 0
+	stride = uint32(skip)
+	for i := 1; i < l; i++ {
+		if v := getBit(bits[i/8], i%8, msb); v != last || stride == maxLite {
+			if err := cb[index].initBlock(physBit, dest[last], stride*4, false, true, false, false, dmaPWM); err != nil {
+				return err
+			}
+			// Hardcoded len(controlBlock) == 32. It is not necessary to use
+			// physToUncachedPhys() here.
+			cb[index].nextCB = uint32(buf.PhysAddr()) + uint32(32*(index+1))
+			index++
+			stride = 0
+			last = v
+		}
+		stride += uint32(skip)
+	}
+	if err := cb[index].initBlock(physBit, dest[last], stride*4, false, true, false, false, dmaPWM); err != nil {
+		return err
+	}
+
+	// Start clock before DMA
+	_, err = setPWMClockSource()
+	if err != nil {
+		return err
+	}
+	return runIO(buf, true)
+}
+
+// dmaWriteStreamDualChannel streams data to a pin using two DMA channels.
+//
+// In practice this leads to a glitchy stream.
+func dmaWriteStreamDualChannel(p *Pin, w gpiostream.Stream) error {
+	// TODO(maruel): Analyse 'w' to figure out the programs to load, and create
+	// the number of controlBlock needed to reduce memory usage.
+	// TODO(maruel): When only one channel is needed, it is much more memory
+	// efficient to use DMA to write to PWM FIFO.
+	skip, err := overSamples(w)
+	if err != nil {
+		return err
+	}
+	l := int(w.Duration()/w.Resolution()) * skip * 4 // Bytes
+	bufLen := (l + 0xFFF) &^ 0xFFF
+	bufSet, err := dmaBufAllocator(bufLen)
+	if err != nil {
+		return err
+	}
+	defer bufSet.Close()
+	bufClear, err := dmaBufAllocator(bufLen)
+	if err != nil {
+		return err
+	}
+	defer bufClear.Close()
+	cb, pCB, err := allocateCB(4096)
+	if err != nil {
+		return err
+	}
+	defer pCB.Close()
+
+	// Needs 64x the memory since each write is 2 full uint32. On the other
+	// hand one could write 32 contiguous pins simultaneously at no cost.
+	mask := uint32(1) << uint(p.number&31)
+	if err := raster32(w, skip, bufClear.Uint32(), bufSet.Uint32(), mask); err != nil {
+		return err
+	}
+
+	// Start clock before DMA start
+	_, err = setPWMClockSource()
+	if err != nil {
+		return err
+	}
+
+	regSet := gpioBaseAddr + 0x1C + 4*uint32(p.number/32)
+	if err := cb[0].initBlock(uint32(bufSet.PhysAddr()), regSet, uint32(l), false, true, true, false, dmaPWM); err != nil {
+		return err
+	}
+	regClear := gpioBaseAddr + 0x28 + 4*uint32(p.number/32)
+	if err := cb[1].initBlock(uint32(bufClear.PhysAddr()), regClear, uint32(l), false, true, true, false, dmaPWM); err != nil {
+		return err
+	}
+
+	// The first channel must be a full bandwidth one.
```

Comment:
@maruel Could you explain why?

Reply:
Because the "light" ones are effectively a single one, which means that they are interleaved. If both are "light" then the jitter is largely increased.

Reply:
I experimented dual channel method but it seems that two channels easily get out of sync. I guess it DREQ clock can be missed.